002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+    - Represents a read-only snapshot of a module's exports.
+    
+### Operations
+    
+* [ParseModule](https://tc39.github.io/ecma262/#sec-parsemodule)
+    - Creates a SourceTextModuleRecord from source code.
+    
+* [HostResolveImportedModule](https://tc39.github.io/ecma262/#sec-hostresolveimportedmodule)
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WhatWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+
+### [Summary Video on Youtube](youtube.com/watch?v=NdOKO-6Ty7k)
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+The list of exports is frozen upon construction. No new properties may be added. No properties may be removed.
+
+## Algorithm
+
+When `require()`ing a file.
+
+1. Determine if file is ES6 or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
```

Comment:
I do not want delusions that they are not compatible. Strict mode code is compatible with sloppy mode code, even though it has different semantics. This falls in the same category.

They have different semantics, but ES6 is a subset that can be completely recreated in CJS, not the opposite. If everyone is wrapped `module.exports` in a Promise, all CJS would only be accessible async. It is also not specified that evaluation is async in the loader spec, as mentioned, just that it finishes evaluation (with top level await this would be async even in CJS, but is complicated and I have not seen public in depth discussion on this problem in spec areas).

Having limits and being a subset can be very valuable for tooling as well as for making guarantees, but it is a super/subset relationship of functionality (not semantics).

>  I don't see any way to avoid this and just return the object; top-level await makes it necessary.

We also are not discussing top level await, it is not anywhere I can dig up outside of that issue really. Perhaps the TC39 discussions can publish something from this week? I would need more public signals as this breaks current run to completion guarantees for both Script and Module evaluation in Ecma262. This is a breakage for existing Module semantics as well.

If we are unable to `require` ES6 modules that means only fully ported code entry points could use ES6 modules. This is an idea of forcing everyone's hand to only write new code in ES6 but you do have restrictions as stated multiple times on this PR as to why that will not always be suitable for all workflows.

If we are unable to `import` CJS we effectively must recreate all modules on NPM from scratch to use them, unacceptable to me and to several others in this PR.